runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+    iree_status_t status = iree_allocator_malloc(
+        tree->allocator, sizeof(iree_tree_node_t) + tree->element_size,
+        (void**)&node);
+    if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+      return status;
+    }
+  }
+  *out_node = node;
+  node->data = (uint8_t*)node + sizeof(iree_tree_node_t);
+  return iree_ok_status();
+}
+
+iree_host_size_t iree_tree_size(iree_tree_t* tree) { return tree->size; }
+
+static bool iree_tree_free_node(iree_tree_node_t* node, void* user_data) {
+  iree_tree_t* tree = (iree_tree_t*)user_data;
+  if ((uint8_t*)node > tree->initial_node_cache &&
+      (uint8_t*)node <
+          tree->initial_node_cache + tree->initial_node_cache_size) {
+    return true;
+  }
+  iree_allocator_free(tree->allocator, node);
+  return true;
+}
+
+void iree_tree_deinitialize(iree_tree_t* tree) {
+  iree_tree_walk(tree, IREE_TREE_WALK_POSTORDER, iree_tree_free_node, tree);
+
+  // Free cache nodes
+  iree_tree_node_t* node = tree->cache;
+  while (node) {
+    iree_tree_node_t* next = node->right;
+    if ((uint8_t*)node < tree->initial_node_cache ||
+        (uint8_t*)node >
+            tree->initial_node_cache + tree->initial_node_cache_size) {
+      iree_allocator_free(tree->allocator, node);
+    }
+    node = next;
+  }
+
+  // Reset the tree structure
+  tree->root = &tree->nil;
+  memset(&tree->nil, 0, sizeof(iree_tree_node_t));
+  tree->nil.is_sentinel = true;
+  tree->size = 0;
+  tree->cache = NULL;
+}
+
+static void iree_tree_rotate_left(iree_tree_t* tree, iree_tree_node_t* x) {
+  iree_tree_node_t* y = x->right;
+  x->right = y->left;
+  if (y->left != &tree->nil) {
+    y->left->parent = x;
+  }
+  y->parent = x->parent;
+  if (x->parent == NULL) {
+    tree->root = y;
+  } else if (x == x->parent->left) {
+    x->parent->left = y;
+  } else {
+    x->parent->right = y;
+  }
+  y->left = x;
+  x->parent = y;
+}
+
+static void iree_tree_rotate_right(iree_tree_t* tree, iree_tree_node_t* x) {
+  iree_tree_node_t* y = x->left;
+  x->left = y->right;
+  if (y->right != &tree->nil) {
+    y->right->parent = x;
+  }
+  y->parent = x->parent;
+  if (x->parent == NULL) {
+    tree->root = y;
+  } else if (x == x->parent->right) {
+    x->parent->right = y;
+  } else {
+    x->parent->left = y;
+  }
+  y->right = x;
+  x->parent = y;
+}
+
+static iree_status_t iree_tree_insert_internal(iree_tree_t* tree,
+                                               iree_host_size_t key,
+                                               iree_tree_node_t* t) {
+  t->left = &tree->nil;
+  t->right = &tree->nil;
+  t->key = key;
+  t->red = true;  // red
+  t->parent = NULL;
+
+  iree_tree_node_t* x = tree->root;
+  iree_tree_node_t* y = NULL;
+  while (x != &tree->nil) {
+    y = x;
+    if (t->key < x->key) {
+      x = x->left;
+    } else if (t->key > x->key) {
+      x = x->right;
+    } else {
+      return iree_make_status(IREE_STATUS_ALREADY_EXISTS,
+                              "Trying to insert a duplicate key");
```

Comment:
mentioned in first review round - please look for all your status messages in the PR and fix them
```suggestion
                              "trying to insert a duplicate key");
```